src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -754,106 +759,252 @@ public static boolean bothSingleByteOptimizable(DynamicObject string, DynamicObj
+        private int countSlow(DynamicObject string, Rope[] ropes, Encoding enc) {
             assert RubyGuards.isRubyString(string);
 
-            DynamicObject otherStr = otherStrings[0];
-            Encoding enc = checkEncodingNode.executeCheckEncoding(string, otherStr);
-
             final boolean[]table = new boolean[StringSupport.TRANS_SIZE + 1];
-            StringSupport.TrTables tables = StringSupport.trSetupTable(rope(otherStr), table, null, true, enc);
-            for (int i = 1; i < otherStrings.length; i++) {
-                otherStr = otherStrings[i];
+            StringSupport.TrTables tables = makeTables(string, ropes, table, enc);
+            return processStr(string, table, enc, tables);
+        }
+    }
 
-                assert RubyGuards.isRubyString(otherStr);
+    public abstract static class TrTableNode extends CoreMethodArrayArgumentsNode {
+        @Child protected ToStrNode toStr = ToStrNodeGen.create(null);
+        @Child protected EncodingNodes.NegotiateCompatibleRopeEncodingNode checkEncodingNode = EncodingNodes.NegotiateCompatibleRopeEncodingNode.create();
 
-                enc = checkEncodingNode.executeCheckEncoding(string, otherStr);
-                tables = StringSupport.trSetupTable(rope(otherStr), table, tables, false, enc);
+        protected boolean[] makeSqueeze() {
+            return new boolean[StringSupport.TRANS_SIZE + 1];
+        }
+
+        protected Encoding findEncoding(DynamicObject string, Rope[] ropes) {
+            Rope rope = StringOperations.rope(string);
+            Encoding enc = checkEncodingNode.executeNegotiate(rope, ropes[0]);
+            for (int i = 1; i < ropes.length; i++) {
+                enc = checkEncodingNode.executeNegotiate(rope, ropes[i]);
+            }
+            return enc;
+        }
+
+        protected TrTables makeTables(DynamicObject string, Rope[] ropes, boolean[] squeeze, Encoding enc) {
+            StringSupport.TrTables tables = StringSupport.trSetupTable(ropes[0], squeeze, null, true, enc);
+
+            for (int i = 1; i < ropes.length; i++) {
+                tables = StringSupport.trSetupTable(ropes[i], squeeze, tables, false, enc);
             }
+            return tables;
+        }
 
-            return StringSupport.strCount(rope(string), table, tables, enc);
+        protected boolean encodingsMatch(DynamicObject string, Encoding encoding) {
+            return encoding == StringOperations.encoding(string);
+        }
+
+        protected Encoding encoding(DynamicObject string) {
+            return StringOperations.encoding(string);
+        }
+
+        @ExplodeLoop
+        protected boolean argsMatch(Rope[] cachedRopes, Rope[] ropes) {
+            if (cachedRopes.length != ropes.length) {
+                return false;
+            } else {
+                for (int i = 0; i < cachedRopes.length; i++) {
+                    if (cachedRopes[i] != ropes[i]) {
```

Comment:
So you want to check not just for equivalent args, but the exact same argument objects? Otherwise, you should use a `RopeNodes.EqualNode`.

Reply:
I guess we can expand it to `RopeNodes.EqualNode`. In general uses of `count`, and friends are extremely static in their arguments (at least all the occurrences I've found are), so we're likely to hit the straight `==` case most of the time. In fact we don't really hit the slow paths in normal code. :-)